### PR TITLE
fix: add cleaning of events attached to dom-events preview node

### PR DESCRIPTION
**Pull request by marcosvega91**
**What**:

Clean up all events attached to preview

**Why**:

Because it can be a problem if in the feature we could jump from events and editor 

**How**:

I have replaced `useCallback` with `useEffect`. On return function of the `useEffect` I have removed events

**Checklist**:

- [ ] Tests
- [X] Ready to be merged

### CODE REVIEW

File: src/components/DomEvents.js [hunk 1]
```diff
@@ -127,18 +132,20 @@ function DomEvents() {
...
       buffer.current.splice(0, 0, event);
       setTimeout(flush, 0);
     });
-  }, []);
+    return () => {
+      eventListeners.forEach((event) =>
+        previewRef.current.removeEventListener(event.name, event.listener),
+      );
+    };
+  }, [previewRef.current]);
```
Comment (smeijer) [inline src/components/DomEvents.js:148]: Will this work? In my experience, `ref.current` values don't (always) trigger `effects`. As the value is mutated. That's why I used the callback method.

So, I think the `removeEventListener` would still be possible, but the `setPreviewRef` callback should get an extra condition:

```js
if (node !== previewRef.current) {
  return;
}

if (previewRef.current) {
  // removeLoggingEvents(previewRef.current, ...
}

if (node) {
  // addLoggingEvents(node, ...
}
```

Comment (smeijer) [reply src/components/DomEvents.js:148]: Source for the above:

https://reactjs.org/docs/hooks-reference.html#useref

> Keep in mind that `useRef` _doesn’t_ notify you when its content changes. Mutating the `.current` property doesn’t cause a re-render. If you want to run some code when React attaches or detaches a ref to a DOM node, you may want to use a callback ref instead.

Comment (marcosvega91) [reply src/components/DomEvents.js:148]: I didn't know this thing but looking around you're right. https://github.com/facebook/react/issues/16121

```diff
diff --git a/src/components/DomEvents.js b/src/components/DomEvents.js
index 7277e8c..47b3660 100644
--- a/src/components/DomEvents.js
+++ b/src/components/DomEvents.js
@@ -119,6 +119,7 @@ function DomEvents() {
   const listRef = useRef();
 
   const [eventCount, setEventCount] = useState(0);
+  const [eventListeners, setEventListeners] = useState([]);
 
   const reset = () => {
     buffer.current = [];
@@ -132,20 +133,23 @@ function DomEvents() {
     [setEventCount],
   );
 
-  useEffect(() => {
-    if (!previewRef.current) return;
-    const eventListeners = addLoggingEvents(previewRef.current, (event) => {
-      // insert at index 0
-      event.id = buffer.current.length;
-      buffer.current.splice(0, 0, event);
-      setTimeout(flush, 0);
-    });
-    return () => {
+  const setPreviewRef = useCallback((node) => {
+    if (node) {
+      previewRef.current = node;
+      const eventListeners = addLoggingEvents(node, (event) => {
+        // insert at index 0
+        event.id = buffer.current.length;
+        buffer.current.splice(0, 0, event);
+        setTimeout(flush, 0);
+      });
+      setEventListeners(eventListeners);
+    } else if (previewRef.current) {
       eventListeners.forEach((event) =>
         previewRef.current.removeEventListener(event.name, event.listener),
       );
-    };
-  }, [previewRef.current]);
+      previewRef.current = null;
+    }
+  }, []);
 
   return (
     <div className="flex flex-col h-auto md:h-full w-full">
@@ -154,7 +158,7 @@ function DomEvents() {
           <MarkupEditor markup={markup} dispatch={dispatch} />
         </div>
 
-        <div className="flex-auto h-56 md:h-full" ref={previewRef}>
+        <div className="flex-auto h-56 md:h-full" ref={setPreviewRef}>
           <Preview
             markup={markup}
             elements={result.elements}
```

File: src/components/DomEvents.js [hunk 2]
```diff
@@ -127,18 +132,20 @@ function DomEvents() {
...
   );
 
-  const setPreviewRef = useCallback((node) => {
-    previewRef.current = node;
-
-    if (!node) return;
-
-    addLoggingEvents(node, (event) => {
+  useEffect(() => {
+    if (!previewRef.current) return;
```
Comment (MichaelDeBoey) [inline src/components/DomEvents.js:136]:
```suggestion
    if (!previewRef.current) {
      return;
    }
```

File: src/components/DomEvents.js [hunk 3]
```diff
@@ -127,18 +132,20 @@ function DomEvents() {
...
       event.id = buffer.current.length;
       buffer.current.splice(0, 0, event);
       setTimeout(flush, 0);
     });
-  }, []);
+    return () => {
+      eventListeners.forEach((event) =>
+        previewRef.current.removeEventListener(event.name, event.listener),
+      );
+    };
```
Comment (MichaelDeBoey) [inline src/components/DomEvents.js:147]:
```suggestion
    return () =>
      eventListeners.forEach((event) =>
        previewRef.current.removeEventListener(event.name, event.listener),
      );
```

File: src/components/DomEvents.js (hunk 2 above)
Comment (smeijer) [reply src/components/DomEvents.js:136]: I think we should add an eslint rule for this one, so it will be auto fixed on commit.

Comment (MichaelDeBoey) [reply src/components/DomEvents.js:136]: Good point!

File: src/components/DomEvents.js (hunk 3 above)
Comment (marcosvega91) [reply src/components/DomEvents.js]: I have changed the implementation

File: src/components/DomEvents.js (hunk 2 above)
Comment (marcosvega91) [reply src/components/DomEvents.js]: I have changed the implementation

File: src/components/DomEvents.js (hunk 1 above)
Comment (marcosvega91) [reply src/components/DomEvents.js]: I have committed following the diff. If you have any doubt let me know 👯

**Pull request merged**